Driver should support nullable types in arrays

### JDBC version in pom.xml
2.2

### ClickHouse version
20.4.4

### Error  logs
```
Caused by: java.sql.SQLException: DataTypeNullable serializeBinary not supported
	at com.github.housepower.jdbc.data.type.complex.DataTypeNullable.serializeBinary(DataTypeNullable.java:68)
	at com.github.housepower.jdbc.data.type.complex.DataTypeArray.serializeBinary(DataTypeArray.java:92)
	at com.github.housepower.jdbc.data.Column.write(Column.java:61)
	at com.github.housepower.jdbc.data.Block.appendRow(Block.java:50)
	at com.github.housepower.jdbc.statement.ClickHousePreparedInsertStatement.addParameters(ClickHousePreparedInsertStatement.java:73)
	at com.github.housepower.jdbc.statement.ClickHousePreparedInsertStatement.addBatch(ClickHousePreparedInsertStatement.java:62)
	at ru.raiffeisen.marketdata.stores.clickhouse.ClickHouseMessageStore.lambda$storeIncrements$5(ClickHouseMessageStore.java:227)
	... 81 more
```
### Steps to reproduce
```SQL
CREATE TABLE increment
(
    asks Nested(price Nullable(Decimal64(9)))
) ENGINE = MergeTree();
```
insert something into table

### Other descriptions

Need to be supported nullable types in arrays

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/ClickHouseResultSet.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/ClickHouseResultSet.java
@@ -15,7 +15,7 @@
 package com.github.housepower.jdbc;
 
 import com.github.housepower.jdbc.data.Block;
-import com.github.housepower.jdbc.data.Column;
+import com.github.housepower.jdbc.data.IColumn;
 import com.github.housepower.jdbc.misc.CheckedIterator;
 import com.github.housepower.jdbc.misc.Validate;
 import com.github.housepower.jdbc.protocol.DataResponse;
@@ -215,7 +215,7 @@ public class ClickHouseResultSet extends SQLResultSet {
     public Object getObject(int index) throws SQLException {
         Validate.isTrue(row >= 0 && row < current.rows(),
             "No row information was obtained.You must call ResultSet.next() before that.");
-        Column column = (lastFetchBlock = current).getByPosition((lastFetchColumn = index - 1));
+        IColumn column = (lastFetchBlock = current).getByPosition((lastFetchColumn = index - 1));
         return column.values((lastFetchRow = row));
     }
 

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/Block.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/Block.java
@@ -35,7 +35,7 @@ public class Block {
         int columns = (int) deserializer.readVarInt();
         int rows = (int) deserializer.readVarInt();
 
-        Column[] cols = new Column[columns];
+        IColumn[] cols = new IColumn[columns];
 
         for (int i = 0; i < columns; i++) {
             String name = deserializer.readStringBinary();
@@ -43,13 +43,13 @@ public class Block {
 
             IDataType dataType = DataTypeFactory.get(type, serverInfo);
             Object[] arr = dataType.deserializeBinaryBulk(rows, deserializer);
-            cols[i] = new Column(name, dataType, arr);
+            cols[i] = ColumnFactory.createColumn(name, dataType, arr);
         }
 
         return new Block(rows, cols, info);
     }
 
-    private final Column[] columns;
+    private final IColumn[] columns;
     private final BlockSettings settings;
     private final Map<String, Integer> nameWithPosition;
 
@@ -58,14 +58,14 @@ public class Block {
     private int rows;
 
     public Block() {
-        this(0, new Column[0]);
+        this(0, new IColumn[0]);
     }
 
-    public Block(int rows, Column[] columns) {
+    public Block(int rows, IColumn[] columns) {
         this(rows, columns, new BlockSettings(Setting.values()));
     }
 
-    public Block(int rows, Column[] columns, BlockSettings settings) {
+    public Block(int rows, IColumn[] columns, BlockSettings settings) {
         this.rows = rows;
         this.columns = columns;
         this.settings = settings;
@@ -111,7 +111,7 @@ public class Block {
         serializer.writeVarInt(columns.length);
         serializer.writeVarInt(rows);
 
-        for (Column column : columns) {
+        for (IColumn column : columns) {
             column.serializeBinaryBulk(serializer);
         }
     }
@@ -124,7 +124,7 @@ public class Block {
         return columns.length;
     }
 
-    public Column getByPosition(int column) throws SQLException {
+    public IColumn getByPosition(int column) throws SQLException {
         Validate.isTrue(column < columns.length,
                 "Position " + column +
                         " is out of bound in Block.getByPosition, max position = " + (columns.length - 1));
@@ -144,7 +144,7 @@ public class Block {
     }
 
     public void initWriteBuffer() {
-        for (Column column : columns) {
+        for (IColumn column : columns) {
             column.initWriteBuffer();
         }
     }

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/Block.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/Block.java
@@ -112,7 +112,7 @@ public class Block {
         serializer.writeVarInt(rows);
 
         for (IColumn column : columns) {
-            column.serializeBinaryBulk(serializer);
+            column.flushToSerializer(serializer, true);
         }
     }
 

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/Block.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/Block.java
@@ -145,7 +145,7 @@ public class Block {
 
     public void initWriteBuffer() {
         for (IColumn column : columns) {
-            column.initWriteBuffer();
+            column.setColumnWriterBuffer(new ColumnWriterBuffer());
         }
     }
 }

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/Column.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/Column.java
@@ -20,24 +20,10 @@ import java.io.IOException;
 import java.sql.SQLException;
 
 public class Column extends AbstractColumn {
-    protected Object[] values;
 
     public Column(String name, IDataType type, Object[] values) {
-        super(name, type);
+        super(name, type, values);
         this.values = values;
-    }
-
-    public String name() {
-        return this.name;
-    }
-
-    public IDataType type() {
-        return this.type;
-    }
-
-    @Override
-    public Object values(int idx) {
-        return values[idx];
     }
 
     @Override
@@ -47,21 +33,10 @@ public class Column extends AbstractColumn {
 
     @Override
     public void serializeBinaryBulk(BinarySerializer serializer) throws SQLException, IOException {
-        serializer.writeStringBinary(name);
-        serializer.writeStringBinary(type.name());
-        buffer.writeTo(serializer);
-    }
-
-    @Override
-    public void clear() {
-        values = new Object[0];
-    }
-
-    public long size() {
-        return values.length;
-    }
-
-    public ColumnWriterBuffer getBuffer() {
-        return buffer;
+        if (isExported()) {
+            serializer.writeStringBinary(name);
+            serializer.writeStringBinary(type.name());
+            buffer.writeTo(serializer);
+        }
     }
 }

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/Column.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/Column.java
@@ -32,10 +32,13 @@ public class Column extends AbstractColumn {
     }
 
     @Override
-    public void serializeBinaryBulk(BinarySerializer serializer) throws SQLException, IOException {
+    public void flushToSerializer(BinarySerializer serializer, boolean now) throws SQLException, IOException {
         if (isExported()) {
             serializer.writeStringBinary(name);
             serializer.writeStringBinary(type.name());
+        }
+
+        if (now) {
             buffer.writeTo(serializer);
         }
     }

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnArray.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnArray.java
@@ -14,6 +14,7 @@
 
 package com.github.housepower.jdbc.data;
 
+import com.github.housepower.jdbc.ClickHouseArray;
 import com.github.housepower.jdbc.data.type.complex.DataTypeArray;
 import com.github.housepower.jdbc.serializer.BinarySerializer;
 
@@ -25,56 +26,51 @@ import java.util.List;
 public class ColumnArray extends AbstractColumn {
     private List<Long> offsets;
     // data represents netsted column in ColumnArray
-    private List<Column> data;
+    private IColumn data;
 
-    public ColumnArray(String name, DataTypeArray type, Object[] values) {
-        super(name, type);
-        offsets = new ArrayList<>(values.length);
-        data = new ArrayList<>(values.length);
-
-        long offset = 0;
-        for (int i = 0; i < values.length; i++) {
-            offsets.set(i, offset);
-            Object []nestData = (Object[])(values[i]);
-            data.set(i, ColumnFactory.createColumn(name, type, nestData));
-
-            offset += nestData.length;
-        }
+    public ColumnArray(String name, DataTypeArray type, Object []values) {
+        super(name, type, values);
+        offsets = new ArrayList<>();
+        data = ColumnFactory.createColumn(null, type.getElemDataType(), null);
     }
 
     @Override
     public void write(Object object) throws IOException, SQLException {
-        type().serializeBinary(object, buffer.column);
+        Object []arr = (Object[]) ((ClickHouseArray) object).getArray();
 
-        Object []arr = (Object[])object;
-        offsets.add(offsets.isEmpty() ? arr.length : offsets.get((int) (size() - 1)) + arr.length);
+        offsets.add(offsets.isEmpty() ? arr.length : offsets.get((offsets.size() - 1)) + arr.length);
+        for (Object field : arr) {
+            data.write(field);
+        }
     }
 
     @Override
     public void serializeBinaryBulk(BinarySerializer serializer) throws SQLException, IOException {
-        serializer.writeStringBinary(name);
-        serializer.writeStringBinary(type.name());
+        if (isExported()) {
+            serializer.writeStringBinary(name);
+            serializer.writeStringBinary(type.name());
+        }
 
         for (long offsetList : offsets) {
             serializer.writeLong(offsetList);
         }
 
-        buffer.writeTo(serializer);
+        data.serializeBinaryBulk(serializer);
+
+        if (isExported()) {
+            buffer.writeTo(serializer);
+        }
+    }
+
+    @Override
+    public void setColumnWriterBuffer(ColumnWriterBuffer buffer) {
+        super.setColumnWriterBuffer(buffer);
+        data.setColumnWriterBuffer(buffer);
     }
 
     @Override
     public void clear() {
         offsets.clear();
         data.clear();
-    }
-
-    @Override
-    public long size() {
-        return 0;
-    }
-
-    @Override
-    public Object values(int idx) {
-        return data.get(idx);
     }
 }

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnArray.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnArray.java
@@ -14,54 +14,67 @@
 
 package com.github.housepower.jdbc.data;
 
+import com.github.housepower.jdbc.data.type.complex.DataTypeArray;
 import com.github.housepower.jdbc.serializer.BinarySerializer;
 
 import java.io.IOException;
 import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
 
-public class Column extends AbstractColumn {
-    protected Object[] values;
+public class ColumnArray extends AbstractColumn {
+    private List<Long> offsets;
+    // data represents netsted column in ColumnArray
+    private List<Column> data;
 
-    public Column(String name, IDataType type, Object[] values) {
+    public ColumnArray(String name, DataTypeArray type, Object[] values) {
         super(name, type);
-        this.values = values;
-    }
+        offsets = new ArrayList<>(values.length);
+        data = new ArrayList<>(values.length);
 
-    public String name() {
-        return this.name;
-    }
+        long offset = 0;
+        for (int i = 0; i < values.length; i++) {
+            offsets.set(i, offset);
+            Object []nestData = (Object[])(values[i]);
+            data.set(i, ColumnFactory.createColumn(name, type, nestData));
 
-    public IDataType type() {
-        return this.type;
-    }
-
-    @Override
-    public Object values(int idx) {
-        return values[idx];
+            offset += nestData.length;
+        }
     }
 
     @Override
     public void write(Object object) throws IOException, SQLException {
         type().serializeBinary(object, buffer.column);
+
+        Object []arr = (Object[])object;
+        offsets.add(offsets.isEmpty() ? arr.length : offsets.get((int) (size() - 1)) + arr.length);
     }
 
     @Override
     public void serializeBinaryBulk(BinarySerializer serializer) throws SQLException, IOException {
         serializer.writeStringBinary(name);
         serializer.writeStringBinary(type.name());
+
+        for (long offsetList : offsets) {
+            serializer.writeLong(offsetList);
+        }
+
         buffer.writeTo(serializer);
     }
 
     @Override
     public void clear() {
-        values = new Object[0];
+        offsets.clear();
+        data.clear();
     }
 
+    @Override
     public long size() {
-        return values.length;
+        return 0;
     }
 
-    public ColumnWriterBuffer getBuffer() {
-        return buffer;
+    @Override
+    public Object values(int idx) {
+        return data.get(idx);
     }
 }

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnFactory.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnFactory.java
@@ -20,11 +20,11 @@ import com.github.housepower.jdbc.data.type.complex.DataTypeNullable;
 import java.sql.Types;
 
 public class ColumnFactory {
-    public static Column createColumn(String name, IDataType type, Object[] values) {
+    public static IColumn createColumn(String name, IDataType type, Object[] values) {
         if (type.sqlTypeId() == Types.ARRAY) {
-            return new ColumnArray(name, (DataTypeArray)type, values);
+            return new ColumnArray(name, (DataTypeArray) type, values);
         } else if (type.nullable()) {
-            return new ColumnNullable(name, (DataTypeNullable)type, values);
+            return new ColumnNullable(name, (DataTypeNullable) type, values);
         }
         return new Column(name, type, values);
     }

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnFactory.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnFactory.java
@@ -1,0 +1,31 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.github.housepower.jdbc.data;
+
+import com.github.housepower.jdbc.data.type.complex.DataTypeArray;
+import com.github.housepower.jdbc.data.type.complex.DataTypeNullable;
+
+import java.sql.Types;
+
+public class ColumnFactory {
+    public static Column createColumn(String name, IDataType type, Object[] values) {
+        if (type.sqlTypeId() == Types.ARRAY) {
+            return new ColumnArray(name, (DataTypeArray)type, values);
+        } else if (type.nullable()) {
+            return new ColumnNullable(name, (DataTypeNullable)type, values);
+        }
+        return new Column(name, type, values);
+    }
+}

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnFactory.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnFactory.java
@@ -16,6 +16,7 @@ package com.github.housepower.jdbc.data;
 
 import com.github.housepower.jdbc.data.type.complex.DataTypeArray;
 import com.github.housepower.jdbc.data.type.complex.DataTypeNullable;
+import com.github.housepower.jdbc.data.type.complex.DataTypeTuple;
 
 import java.sql.Types;
 
@@ -25,6 +26,8 @@ public class ColumnFactory {
             return new ColumnArray(name, (DataTypeArray) type, values);
         } else if (type.nullable()) {
             return new ColumnNullable(name, (DataTypeNullable) type, values);
+        } else if (type.sqlTypeId() == Types.STRUCT) {
+            return new ColumnTuple(name, (DataTypeTuple) type, values);
         }
         return new Column(name, type, values);
     }

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnNullable.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnNullable.java
@@ -40,7 +40,7 @@ public class ColumnNullable extends AbstractColumn {
     }
 
     @Override
-    public void serializeBinaryBulk(BinarySerializer serializer) throws SQLException, IOException {
+    public void flushToSerializer(BinarySerializer serializer, boolean now) throws SQLException, IOException {
         if (isExported()) {
             serializer.writeStringBinary(name);
             serializer.writeStringBinary(type.name());
@@ -50,9 +50,8 @@ public class ColumnNullable extends AbstractColumn {
             serializer.writeByte(sign);
         }
 
-        if (isExported()) {
+        if (now)
             buffer.writeTo(serializer);
-        }
     }
 
     @Override

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnNullable.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnNullable.java
@@ -22,43 +22,42 @@ import java.sql.SQLException;
 import java.util.ArrayList;
 import java.util.List;
 
-public class ColumnNullable extends Column{
+public class ColumnNullable extends AbstractColumn {
     private List<Byte> nullableSign;
     // data represents netsted column in ColumnArray
-    private Column data;
+    private IColumn data;
 
-    public ColumnNullable(String name, DataTypeNullable type, Object[] values) {
+    public ColumnNullable(String name, DataTypeNullable type, Object []values) {
         super(name, type, values);
-        nullableSign = new ArrayList<>(values.length);
-        for (int i = 0; i < values.length; i++) {
-            nullableSign.set(i, values[i] == null ? (byte) 1 : 0);
-        }
-        data = ColumnFactory.createColumn(name, type, values);
+        nullableSign = new ArrayList<>();
+        data = ColumnFactory.createColumn(null, type.getNestedDataType(), null);
     }
 
     @Override
     public void write(Object object) throws IOException, SQLException {
         nullableSign.add(object == null ? (byte) 1 : 0);
-        data.write(object);
+        data.write(object == null ? type.defaultValue() : object);
     }
 
     @Override
     public void serializeBinaryBulk(BinarySerializer serializer) throws SQLException, IOException {
-        serializer.writeStringBinary(name);
-        serializer.writeStringBinary(type.name());
+        if (isExported()) {
+            serializer.writeStringBinary(name);
+            serializer.writeStringBinary(type.name());
+        }
 
         for (byte sign : nullableSign) {
             serializer.writeByte(sign);
         }
 
-        data.serializeBinaryBulk(serializer);
-        buffer.writeTo(serializer);
+        if (isExported()) {
+            buffer.writeTo(serializer);
+        }
     }
 
     @Override
-    public void clear() {
-        super.clear();
-        nullableSign.clear();
-        data.clear();
+    public void setColumnWriterBuffer(ColumnWriterBuffer buffer) {
+        super.setColumnWriterBuffer(buffer);
+        data.setColumnWriterBuffer(buffer);
     }
 }

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnNullable.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnNullable.java
@@ -14,54 +14,51 @@
 
 package com.github.housepower.jdbc.data;
 
+import com.github.housepower.jdbc.data.type.complex.DataTypeNullable;
 import com.github.housepower.jdbc.serializer.BinarySerializer;
 
 import java.io.IOException;
 import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
 
-public class Column extends AbstractColumn {
-    protected Object[] values;
+public class ColumnNullable extends Column{
+    private List<Byte> nullableSign;
+    // data represents netsted column in ColumnArray
+    private Column data;
 
-    public Column(String name, IDataType type, Object[] values) {
-        super(name, type);
-        this.values = values;
-    }
-
-    public String name() {
-        return this.name;
-    }
-
-    public IDataType type() {
-        return this.type;
-    }
-
-    @Override
-    public Object values(int idx) {
-        return values[idx];
+    public ColumnNullable(String name, DataTypeNullable type, Object[] values) {
+        super(name, type, values);
+        nullableSign = new ArrayList<>(values.length);
+        for (int i = 0; i < values.length; i++) {
+            nullableSign.set(i, values[i] == null ? (byte) 1 : 0);
+        }
+        data = ColumnFactory.createColumn(name, type, values);
     }
 
     @Override
     public void write(Object object) throws IOException, SQLException {
-        type().serializeBinary(object, buffer.column);
+        nullableSign.add(object == null ? (byte) 1 : 0);
+        data.write(object);
     }
 
     @Override
     public void serializeBinaryBulk(BinarySerializer serializer) throws SQLException, IOException {
         serializer.writeStringBinary(name);
         serializer.writeStringBinary(type.name());
+
+        for (byte sign : nullableSign) {
+            serializer.writeByte(sign);
+        }
+
+        data.serializeBinaryBulk(serializer);
         buffer.writeTo(serializer);
     }
 
     @Override
     public void clear() {
-        values = new Object[0];
-    }
-
-    public long size() {
-        return values.length;
-    }
-
-    public ColumnWriterBuffer getBuffer() {
-        return buffer;
+        super.clear();
+        nullableSign.clear();
+        data.clear();
     }
 }

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnTuple.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/ColumnTuple.java
@@ -14,33 +14,32 @@
 
 package com.github.housepower.jdbc.data;
 
-import com.github.housepower.jdbc.ClickHouseArray;
-import com.github.housepower.jdbc.data.type.complex.DataTypeArray;
+import com.github.housepower.jdbc.ClickHouseStruct;
+import com.github.housepower.jdbc.data.type.complex.DataTypeTuple;
 import com.github.housepower.jdbc.serializer.BinarySerializer;
 
 import java.io.IOException;
 import java.sql.SQLException;
-import java.util.ArrayList;
-import java.util.List;
 
-public class ColumnArray extends AbstractColumn {
-    private List<Long> offsets;
+public class ColumnTuple extends AbstractColumn {
     // data represents netsted column in ColumnArray
-    private IColumn data;
+    private IColumn []datas;
 
-    public ColumnArray(String name, DataTypeArray type, Object []values) {
+    public ColumnTuple(String name, DataTypeTuple type, Object []values) {
         super(name, type, values);
-        offsets = new ArrayList<>();
-        data = ColumnFactory.createColumn(null, type.getElemDataType(), null);
+
+        IDataType []types = type.getNestedTypes();
+        datas = new IColumn[types.length];
+        for (int i = 0; i < types.length; i++) {
+            datas[i] = ColumnFactory.createColumn(null, types[i], null);
+        }
     }
 
     @Override
     public void write(Object object) throws IOException, SQLException {
-        Object []arr = (Object[]) ((ClickHouseArray) object).getArray();
-
-        offsets.add(offsets.isEmpty() ? arr.length : offsets.get((offsets.size() - 1)) + arr.length);
-        for (Object field : arr) {
-            data.write(field);
+        ClickHouseStruct tuple = (ClickHouseStruct) object;
+        for (int i = 0; i < datas.length; i ++) {
+            datas[i].write(tuple.getAttributes()[i]);
         }
     }
 
@@ -51,29 +50,27 @@ public class ColumnArray extends AbstractColumn {
             serializer.writeStringBinary(type.name());
         }
 
-        flushOffsets(serializer);
-        data.flushToSerializer(serializer, false);
+        // we should to flush all the nested data to serializer
+        // because they are using separate buffers.
+        for (int i = 0; i < datas.length; i ++) {
+            datas[i].flushToSerializer(serializer, true);
+        }
 
         if (now) {
             buffer.writeTo(serializer);
         }
     }
 
-    public void flushOffsets(BinarySerializer serializer) throws SQLException, IOException {
-        for (long offsetList : offsets) {
-            serializer.writeLong(offsetList);
+    @Override
+    public void setColumnWriterBuffer(ColumnWriterBuffer buffer) {
+        super.setColumnWriterBuffer(buffer);
+
+        for (int i = 0; i < datas.length; i ++) {
+            datas[i].setColumnWriterBuffer(new ColumnWriterBuffer());
         }
     }
 
     @Override
-    public void setColumnWriterBuffer(ColumnWriterBuffer buffer) {
-        super.setColumnWriterBuffer(buffer);
-        data.setColumnWriterBuffer(buffer);
-    }
-
-    @Override
     public void clear() {
-        offsets.clear();
-        data.clear();
     }
 }

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/IColumn.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/IColumn.java
@@ -23,12 +23,12 @@ public interface IColumn {
     void write(Object object) throws IOException, SQLException;
     void serializeBinaryBulk(BinarySerializer serializer) throws SQLException, IOException;
     void clear();
-    long size();
 
-    void initWriteBuffer();
+    void setColumnWriterBuffer(ColumnWriterBuffer buffer);
 
     String name();
     IDataType type();
+    boolean isExported();
     Object values(int idx);
 }
 
@@ -36,11 +36,19 @@ public interface IColumn {
 abstract class AbstractColumn implements IColumn {
     protected final String name;
     protected final IDataType type;
+
+    protected Object []values;
     protected ColumnWriterBuffer buffer;
 
-    public AbstractColumn(String name, IDataType type) {
+    public AbstractColumn(String name, IDataType type, Object []values) {
         this.name = name;
         this.type = type;
+        this.values = values;
+    }
+
+    @Override
+    public boolean isExported() {
+        return name != null;
     }
 
     @Override
@@ -54,7 +62,17 @@ abstract class AbstractColumn implements IColumn {
     }
 
     @Override
-    public void initWriteBuffer() {
-        this.buffer = new ColumnWriterBuffer();
+    public Object values(int idx) {
+        return values[idx];
+    }
+
+    @Override
+    public void clear() {
+        values = new Object[0];
+    }
+
+    @Override
+    public void setColumnWriterBuffer(ColumnWriterBuffer buffer) {
+        this.buffer = buffer;
     }
 }

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/IColumn.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/IColumn.java
@@ -37,6 +37,7 @@ abstract class AbstractColumn implements IColumn {
     protected final String name;
     protected final IDataType type;
 
+    // Note: values is only for reading
     protected Object []values;
     protected ColumnWriterBuffer buffer;
 

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/IColumn.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/IColumn.java
@@ -1,0 +1,60 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.github.housepower.jdbc.data;
+
+import com.github.housepower.jdbc.serializer.BinarySerializer;
+
+import java.io.IOException;
+import java.sql.SQLException;
+
+public interface IColumn {
+    void write(Object object) throws IOException, SQLException;
+    void serializeBinaryBulk(BinarySerializer serializer) throws SQLException, IOException;
+    void clear();
+    long size();
+
+    void initWriteBuffer();
+
+    String name();
+    IDataType type();
+    Object values(int idx);
+}
+
+
+abstract class AbstractColumn implements IColumn {
+    protected final String name;
+    protected final IDataType type;
+    protected ColumnWriterBuffer buffer;
+
+    public AbstractColumn(String name, IDataType type) {
+        this.name = name;
+        this.type = type;
+    }
+
+    @Override
+    public String name() {
+        return name;
+    }
+
+    @Override
+    public IDataType type() {
+        return type;
+    }
+
+    @Override
+    public void initWriteBuffer() {
+        this.buffer = new ColumnWriterBuffer();
+    }
+}

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/IColumn.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/IColumn.java
@@ -21,10 +21,19 @@ import java.sql.SQLException;
 
 public interface IColumn {
     void write(Object object) throws IOException, SQLException;
-    void serializeBinaryBulk(BinarySerializer serializer) throws SQLException, IOException;
+
+    /**
+     *
+     * @param serializer is serializer wrapper of tcp socket
+     * @param now means we should flush all the buffer to serializer now
+     * @throws SQLException
+     * @throws IOException
+     */
+    void flushToSerializer(BinarySerializer serializer, boolean now) throws SQLException, IOException;
     void clear();
 
     void setColumnWriterBuffer(ColumnWriterBuffer buffer);
+    ColumnWriterBuffer getColumnWriterBuffer();
 
     String name();
     IDataType type();
@@ -75,5 +84,10 @@ abstract class AbstractColumn implements IColumn {
     @Override
     public void setColumnWriterBuffer(ColumnWriterBuffer buffer) {
         this.buffer = buffer;
+    }
+
+    @Override
+    public ColumnWriterBuffer getColumnWriterBuffer() {
+        return buffer;
     }
 }

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/IDataType.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/IDataType.java
@@ -20,7 +20,6 @@ import com.github.housepower.jdbc.serializer.BinarySerializer;
 
 import java.io.IOException;
 import java.sql.SQLException;
-import java.util.Iterator;
 
 // It would be nice if we introduce a Generic Type, `IDataType<T>`, then we can avoid using `Object` and type cast.
 // Unfortunately Java not support unsigned number, UInt8(u_byte) must be represented by Int16(short), which will
@@ -46,12 +45,6 @@ public interface IDataType {
     Object deserializeBinary(BinaryDeserializer deserializer) throws SQLException, IOException;
 
     void serializeBinary(Object data, BinarySerializer serializer) throws SQLException, IOException;
-
-    default void serializeBinaryBulk(Iterator<Object> data, BinarySerializer serializer) throws SQLException, IOException {
-        while (data.hasNext()) {
-            serializeBinary(data.next(), serializer);
-        }
-    }
 
     default void serializeBinaryBulk(Object[] data, BinarySerializer serializer) throws SQLException, IOException {
         for (Object d : data) {

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/type/complex/DataTypeArray.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/type/complex/DataTypeArray.java
@@ -43,6 +43,15 @@ public class DataTypeArray implements IDataType {
 
     private final String name;
     private final Array defaultValue;
+
+    public IDataType getElemDataType() {
+        return elemDataType;
+    }
+
+    public IDataType getOffsetIDataType() {
+        return offsetIDataType;
+    }
+
     private final IDataType elemDataType;
     private final IDataType offsetIDataType;
 

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/type/complex/DataTypeArray.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/type/complex/DataTypeArray.java
@@ -29,7 +29,6 @@ import java.sql.Array;
 import java.sql.SQLException;
 import java.sql.Types;
 import java.util.ArrayList;
-import java.util.Iterator;
 import java.util.List;
 
 public class DataTypeArray implements IDataType {
@@ -132,14 +131,10 @@ public class DataTypeArray implements IDataType {
 
     @Override
     public void serializeBinary(Object data, BinarySerializer serializer) throws SQLException, IOException {
+
         throw new SQLException("DataTypeArray serializeBinary not supported");
     }
 
-    @Override
-    public void serializeBinaryBulk(Iterator<Object> data, BinarySerializer serializer)
-            throws SQLException, IOException {
-        throw new SQLException("DataTypeArray serializeBinaryBulk not supported");
-    }
 
     @Override
     public void serializeBinaryBulk(Object[] data, BinarySerializer serializer)

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/type/complex/DataTypeArray.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/type/complex/DataTypeArray.java
@@ -48,10 +48,6 @@ public class DataTypeArray implements IDataType {
         return elemDataType;
     }
 
-    public IDataType getOffsetIDataType() {
-        return offsetIDataType;
-    }
-
     private final IDataType elemDataType;
     private final IDataType offsetIDataType;
 
@@ -114,41 +110,20 @@ public class DataTypeArray implements IDataType {
         return new ClickHouseArray(arrayData.toArray());
     }
 
-    public void serializeBinary(Object data,
-                                BinarySerializer dataBinarySerializer,
-                                List<List<Integer>> offsets,
-                                int level) throws SQLException, IOException {
-        int dataOffset = ((Object[]) ((Array) data).getArray()).length;
-        if (offsets.size() < level) {
-            List<Integer> offset = new ArrayList<>();
-            offset.add(dataOffset);
-            offsets.add(offset);
-        } else {
-            int lastIdx = offsets.get(level - 1).size() - 1;
-            int lastOffset = offsets.get(level - 1).get(lastIdx);
-            offsets.get(level - 1).add(lastOffset + dataOffset);
-        }
-
-        for (Object v : ((Object[]) ((Array) data).getArray())) {
-            if (elemDataType.sqlTypeId() == Types.ARRAY) {
-                ((DataTypeArray) (elemDataType)).serializeBinary(v, dataBinarySerializer, offsets, level + 1);
-            } else {
-                elemDataType.serializeBinary(v, dataBinarySerializer);
-            }
-        }
-    }
-
     @Override
     public void serializeBinary(Object data, BinarySerializer serializer) throws SQLException, IOException {
-
-        throw new SQLException("DataTypeArray serializeBinary not supported");
+        Object []arr = (Object[]) data;
+        for (Object f : arr) {
+            getElemDataType().serializeBinary(f, serializer);
+        }
     }
 
 
     @Override
     public void serializeBinaryBulk(Object[] data, BinarySerializer serializer)
             throws SQLException, IOException {
-        throw new SQLException("DataTypeArray serializeBinaryBulk not supported");
+        offsetIDataType.serializeBinary(data.length, serializer);
+        getElemDataType().serializeBinaryBulk(data, serializer);
     }
 
     @Override

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/type/complex/DataTypeNullable.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/type/complex/DataTypeNullable.java
@@ -41,6 +41,11 @@ public class DataTypeNullable implements IDataType {
 
     private final String name;
     private final IDataType nestedDataType;
+
+    public IDataType getNestedDataType() {
+        return nestedDataType;
+    }
+
     private final IDataType nullMapDataType;
 
     public DataTypeNullable(String name, IDataType nestedDataType, IDataType nullMapIDataType) throws SQLException {
@@ -98,12 +103,7 @@ public class DataTypeNullable implements IDataType {
 
     @Override
     public void serializeBinary(Object data, BinarySerializer serializer) throws SQLException, IOException {
-        if (data == null) {
-            serializer.writeByte((byte) 1);
-        } else {
-            serializer.writeByte((byte) 0);
-            this.nestedDataType.serializeBinary(data, serializer);
-        }
+        this.nestedDataType.serializeBinary(data, serializer);
     }
 
     public void serializeBinary(Object data, BinarySerializer serializer, List<Byte> offset) throws SQLException, IOException {
@@ -113,7 +113,7 @@ public class DataTypeNullable implements IDataType {
 
     @Override
     public Object deserializeBinary(BinaryDeserializer deserializer) throws SQLException, IOException {
-        boolean isNull = (deserializer.readByte() == (byte)1);
+        boolean isNull = (deserializer.readByte() == (byte) 1);
         if (isNull) {
             return null;
         }

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/type/complex/DataTypeNullable.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/type/complex/DataTypeNullable.java
@@ -24,7 +24,6 @@ import com.github.housepower.jdbc.serializer.BinarySerializer;
 
 import java.io.IOException;
 import java.sql.SQLException;
-import java.util.List;
 
 public class DataTypeNullable implements IDataType {
 
@@ -104,11 +103,6 @@ public class DataTypeNullable implements IDataType {
     @Override
     public void serializeBinary(Object data, BinarySerializer serializer) throws SQLException, IOException {
         this.nestedDataType.serializeBinary(data, serializer);
-    }
-
-    public void serializeBinary(Object data, BinarySerializer serializer, List<Byte> offset) throws SQLException, IOException {
-        offset.add(data == null ? (byte) 1 : 0);
-        this.nestedDataType.serializeBinary(data == null ? nestedDataType.defaultValue() : data, serializer);
     }
 
     @Override

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/type/complex/DataTypeNullable.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/type/complex/DataTypeNullable.java
@@ -98,7 +98,12 @@ public class DataTypeNullable implements IDataType {
 
     @Override
     public void serializeBinary(Object data, BinarySerializer serializer) throws SQLException, IOException {
-        throw new SQLException("DataTypeNullable serializeBinary not supported");
+        if (data == null) {
+            serializer.writeByte((byte) 1);
+        } else {
+            serializer.writeByte((byte) 0);
+            this.nestedDataType.serializeBinary(data, serializer);
+        }
     }
 
     public void serializeBinary(Object data, BinarySerializer serializer, List<Byte> offset) throws SQLException, IOException {
@@ -108,7 +113,11 @@ public class DataTypeNullable implements IDataType {
 
     @Override
     public Object deserializeBinary(BinaryDeserializer deserializer) throws SQLException, IOException {
-        throw new SQLException("DataTypeNullable deserializeBinary not supported");
+        boolean isNull = (deserializer.readByte() == (byte)1);
+        if (isNull) {
+            return null;
+        }
+        return this.nestedDataType.deserializeBinary(deserializer);
     }
 
     @Override

--- a/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/type/complex/DataTypeTuple.java
+++ b/clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/data/type/complex/DataTypeTuple.java
@@ -53,6 +53,10 @@ public class DataTypeTuple implements IDataType {
         }
     }
 
+    public IDataType[] getNestedTypes() {
+        return nestedTypes;
+    }
+
     private final String name;
     private final IDataType[] nestedTypes;
 

--- a/clickhouse-native-jdbc/src/test/java/com/github/housepower/jdbc/BatchInsertITest.java
+++ b/clickhouse-native-jdbc/src/test/java/com/github/housepower/jdbc/BatchInsertITest.java
@@ -155,16 +155,18 @@ public class BatchInsertITest extends AbstractITest {
 
             statement.execute("DROP TABLE IF EXISTS test");
             statement.execute(
-                    "CREATE TABLE test(name Array(String), value Array(Float64), value2 Array(Array(Int32)))ENGINE=Log");
+                    "CREATE TABLE test(value0 Array(String), value1 Array(Float64), value2 Array(Array(Int32)), array3 Array(Nullable(Float64)))ENGINE=Log");
             PreparedStatement preparedStatement = connection
-                    .prepareStatement("INSERT INTO test VALUES(?, ?, [[1,2,3]])");
+                    .prepareStatement("INSERT INTO test VALUES(?, ?, [[1,2,3]], ?)");
 
-            List<String> array = Arrays.asList("aa", "bb", "cc");
-            List<Double> array2 = Arrays.asList(1.2, 2.2, 3.2);
+            List<String> array0 = Arrays.asList("aa", "bb", "cc");
+            List<Double> array1 = Arrays.asList(1.2, 2.2, 3.2);
+            List<Double> array3 = Arrays.asList(1.2, 2.2, 3.2, null);
 
             for (int i = 0; i < Byte.MAX_VALUE; i++) {
-                preparedStatement.setArray(1, connection.createArrayOf("text", array.toArray()));
-                preparedStatement.setArray(2, connection.createArrayOf("text", array2.toArray()));
+                preparedStatement.setArray(1, connection.createArrayOf("text", array0.toArray()));
+                preparedStatement.setArray(2, connection.createArrayOf("text", array1.toArray()));
+                preparedStatement.setArray(3, connection.createArrayOf("text", array3.toArray()));
 
                 preparedStatement.addBatch();
             }
@@ -173,8 +175,9 @@ public class BatchInsertITest extends AbstractITest {
 
             ResultSet rs = statement.executeQuery("select * from test");
             while (rs.next()) {
-                assertArrayEquals(array.toArray(), (Object[]) rs.getArray(1).getArray());
-                assertArrayEquals(array2.toArray(), (Object[]) rs.getArray(2).getArray());
+                assertArrayEquals(array0.toArray(), (Object[]) rs.getArray(1).getArray());
+                assertArrayEquals(array1.toArray(), (Object[]) rs.getArray(2).getArray());
+                assertArrayEquals(array3.toArray(), (Object[]) rs.getArray(4).getArray());
             }
         });
 

--- a/clickhouse-native-jdbc/src/test/java/com/github/housepower/jdbc/InsertComplexTypeITest.java
+++ b/clickhouse-native-jdbc/src/test/java/com/github/housepower/jdbc/InsertComplexTypeITest.java
@@ -102,12 +102,12 @@ public class InsertComplexTypeITest extends AbstractITest {
             assertTrue(rs.next());
 
             assertEquals(
-                    Timestamp.valueOf(LocalDateTime.of(2000, 1, 1, 8, 1, 1, 0)).getTime(),
-                    rs.getTimestamp(1).getTime());
+                Timestamp.valueOf(LocalDateTime.of(2000, 1, 1, 8, 1, 1, 0)).getTime(),
+                rs.getTimestamp(1).getTime());
 
             assertEquals(
-                    Timestamp.valueOf(LocalDateTime.of(2000, 1, 1, 8, 1, 1, 0)).getTime(),
-                    rs.getTimestamp(2).getTime());
+                Timestamp.valueOf(LocalDateTime.of(2000, 1, 1, 8, 1, 1, 0)).getTime(),
+                rs.getTimestamp(2).getTime());
 
             assertFalse(rs.next());
             statement.executeQuery("DROP TABLE IF EXISTS test");
@@ -126,8 +126,8 @@ public class InsertComplexTypeITest extends AbstractITest {
             assertTrue(rs.next());
 
             assertEquals(
-                    Timestamp.valueOf(LocalDateTime.of(2000, 1, 1, 0, 1, 1, 123456789)),
-                    rs.getTimestamp(1));
+                Timestamp.valueOf(LocalDateTime.of(2000, 1, 1, 0, 1, 1, 123456789)),
+                rs.getTimestamp(1));
             assertFalse(rs.next());
             statement.executeQuery("DROP TABLE IF EXISTS test");
         }, true);
@@ -144,8 +144,8 @@ public class InsertComplexTypeITest extends AbstractITest {
             ResultSet rs = statement.executeQuery("SELECT * FROM test");
             assertTrue(rs.next());
             assertEquals(
-                    Timestamp.valueOf(LocalDateTime.of(1970, 1, 1, 0, 0, 0, 0)),
-                    rs.getTimestamp(1));
+                Timestamp.valueOf(LocalDateTime.of(1970, 1, 1, 0, 0, 0, 0)),
+                rs.getTimestamp(1));
             assertFalse(rs.next());
             statement.executeQuery("DROP TABLE IF EXISTS test");
         }, true);
@@ -163,8 +163,8 @@ public class InsertComplexTypeITest extends AbstractITest {
             assertTrue(rs.next());
 
             assertEquals(
-                    Timestamp.valueOf(LocalDateTime.of(2105, 12, 31, 23, 59, 59, 999999999)),
-                    rs.getTimestamp(1));
+                Timestamp.valueOf(LocalDateTime.of(2105, 12, 31, 23, 59, 59, 999999999)),
+                rs.getTimestamp(1));
             assertFalse(rs.next());
             statement.executeQuery("DROP TABLE IF EXISTS test");
         }, true);
@@ -176,13 +176,30 @@ public class InsertComplexTypeITest extends AbstractITest {
             Statement statement = connection.createStatement();
 
             statement.executeQuery("DROP TABLE IF EXISTS test");
-            statement.executeQuery("CREATE TABLE test(test_tuple Tuple(String, UInt8))ENGINE=Log");
-            statement.executeQuery("INSERT INTO test VALUES(('test_string', 1))");
+            statement.executeQuery("CREATE TABLE test(test_tuple Tuple(String, UInt8),"
+                                   + " tuple_array  Tuple(Array(Nullable(String)), Nullable(UInt8)),"
+                                   + " array_tuple Array(Tuple(UInt32, Nullable(String)) )"
+                                   + " )ENGINE=Log");
+            statement.executeQuery("INSERT INTO test VALUES( ('test_string', 1), (['1'], 32), [(32, '1'), (22, NULL) ] )");
             ResultSet rs = statement.executeQuery("SELECT * FROM test");
             assertTrue(rs.next());
             assertArrayEquals(
-                    new Object[]{"test_string", (short) (1)},
-                    ((Struct) rs.getObject(1)).getAttributes());
+                new Object[]{"test_string", (short) (1)},
+                ((Struct) rs.getObject(1)).getAttributes());
+
+            Object []objs = ((Struct) rs.getObject(2)).getAttributes();
+
+            ClickHouseArray arr = (ClickHouseArray) (objs[0]);
+            assertArrayEquals(new Object[]{"1"}, (Object[]) arr.getArray());
+            assertEquals((short) 32, objs[1]);
+
+            arr = (ClickHouseArray) rs.getObject(3);
+
+            ClickHouseStruct t1 = (ClickHouseStruct) ((Object[]) arr.getArray())[0];
+            ClickHouseStruct t2 = (ClickHouseStruct) ((Object[]) arr.getArray())[1];
+            assertArrayEquals(new Object[]{(long) 32, "1"}, t1.getAttributes());
+            assertArrayEquals(new Object[]{(long) 22, null}, t2.getAttributes());
+
             assertFalse(rs.next());
             statement.executeQuery("DROP TABLE IF EXISTS test");
         });

--- a/clickhouse-native-jdbc/src/test/java/com/github/housepower/jdbc/InsertComplexTypeITest.java
+++ b/clickhouse-native-jdbc/src/test/java/com/github/housepower/jdbc/InsertComplexTypeITest.java
@@ -29,14 +29,18 @@ public class InsertComplexTypeITest extends AbstractITest {
             Statement statement = connection.createStatement();
 
             statement.executeQuery("DROP TABLE IF EXISTS test");
-            statement.executeQuery("CREATE TABLE test(test_Array Array(UInt8), test_Array2 Array(Array(String)))ENGINE=Log");
-            statement.executeQuery("INSERT INTO test VALUES ([1, 2, 3, 4], [ ['1', '2'] ])");
+            statement.executeQuery("CREATE TABLE test(test_Array Array(UInt8), test_Array2 Array(Array(String)), n3 Array(Nullable(UInt8)) )ENGINE=Log");
+            statement.executeQuery("INSERT INTO test VALUES ([1, 2, 3, 4], [ ['1', '2'] ], [1, 2, NULL] )");
             ResultSet rs = statement.executeQuery("SELECT * FROM test");
             assertTrue(rs.next());
             assertArrayEquals(new Short[]{1, 2, 3, 4}, (Object[]) rs.getArray(1).getArray());
             Object[] objects = (Object[]) rs.getArray(2).getArray();
             ClickHouseArray array = (ClickHouseArray) objects[0];
             assertArrayEquals(new Object[]{"1", "2"}, (Object[]) array.getArray());
+
+            objects = (Object[]) rs.getArray(3).getArray();
+            assertArrayEquals(new Short[]{1, 2, null}, objects);
+
             assertFalse(rs.next());
             statement.executeQuery("DROP TABLE IF EXISTS test");
         });

--- a/clickhouse-native-jdbc/src/test/java/com/github/housepower/jdbc/QueryComplexTypeITest.java
+++ b/clickhouse-native-jdbc/src/test/java/com/github/housepower/jdbc/QueryComplexTypeITest.java
@@ -199,7 +199,7 @@ public class QueryComplexTypeITest extends AbstractITest {
             Statement statement = connection.createStatement();
 
             ResultSet rs = statement.executeQuery(
-                    "SELECT [[1.1, 1.2], [2.1, 2.2], [3.1, 3.2]] AS v, toTypeName(v) from numbers(10)");
+                    "SELECT [[1.1, 1.2], [2.1, 2.2], [3.1, 3.2]] AS v, toTypeName(v), [toNullable(10000), toNullable(10001)] from numbers(10)");
 
             for (int i = 0; i < 10; i++) {
                 assertTrue(rs.next());
@@ -213,6 +213,10 @@ public class QueryComplexTypeITest extends AbstractITest {
                 assertArrayEquals(res[1], (Object[]) ((ClickHouseArray) (arr[1])).getArray());
                 assertArrayEquals(res[2], (Object[]) ((ClickHouseArray) (arr[2])).getArray());
                 assertEquals("Array(Array(Float64))", rs.getString(2));
+
+                arr = (Object[]) (rs.getArray(3).getArray());
+                assertEquals(10000, arr[0]);
+                assertEquals(10001, arr[1]);
             }
             assertFalse(rs.next());
 

--- a/docs/Process.md
+++ b/docs/Process.md
@@ -1,0 +1,54 @@
+ClickHouse client --- server request response process
+========================================
+
+## Connection
+
+ ```sequence
+Client -> Server: Open socket Connection
+Server --> Client: Ok, got a new client connection
+Note right of Server: Connection established
+Client -> Server: Send Hello Request
+Server --> Client: Hello response
+Note left of Client: I got server infos
+ ```
+
+
+## Request
+
+- There are many kinds of requests/response, the above `hello` is one of them.
+
+- You can find all the request/response type in `com.github.housepower.jdbc.protocol` package.
+
+## Data Querys
+
+After the connection established and hello request/response, we can send plain sql strings to query the data. 
+
+ ```sequence
+
+Client -> Server: Send DataRequest Request
+Note right of Server: Oh, a new query just comes, I will handle that query.
+Server --> Client: DataResponse
+Note left of Client: I got response data now
+Note left of Client: I will deserialize them to the ResultSets.
+ ```
+
+ ## Insert Querys
+
+ The plain sql querys which send query sql to the server, but it's not efficient for batch inserts. ClickHouse provide another type of data request for batch inserts that we can send blocks to the server directly.
+
+  ```sequence
+Client -> Server: Send insert query to the server (which called by PreparedStatement)
+Note right of Server: Oh, a new prepare insert just comes, I'll look at the table schemas.
+Server --> Client: DataResponse (Empty Block, which is also called sampleBlock)
+Note right of Server: State: Waiting for inserts.
+Note left of Client: I got a block now, and I know the names and types of this table.
+Note left of Client: Write the data to the blocks (when we can `setObject` in JDBC)
+Client -> Server: send a large block by dataRequest
+Note right of Server: A Block just comes, I'll insert them to the table
+Client -> Server: send a empty block to end the inserts
+Note right of Server: A empty block just comes, which means the client finish the inserts.
+Note right of Server: State: Idle.
+  ```
+
+ 
+

